#  Analysis of Mobile Phone Cost

**Desired outcome **
* In this project we will look at mobile phone cost from the dataset data.csv.
* The idea is to *analyze* the data set, make *conjectures*, support or refute those conjectures with *data*, and *tell a story* about the data!

**Data Description.**

The dataset data.csv is available from our project website: https://github.com/ds3010s22/ds3010_projects, and 
is originally from Kaggle: https://www.kaggle.com/code/dansbecker/classification/data?select=train.csv

The data contains 20 features including 
* battery_power: total energy a battery can store in one time measured in mAh
* blue: has bluetooth or not 
* clock_speed: speed at which microprocessor executes instructions 
* dual_sim: has dual sim support or not
* four_g: has 4G or not
* etc.

The target is the price range, which has 4 classes including 0 (low cost), 1 (medium cost), 2 (high cost) and 3 (very high cost).

**Required Readings:** 
* This project will be based upon the scikit-learn Python library
* Read about deep learning at https://scikit-learn.org/stable/modules/neural_networks_supervised.html
* Read about Scikit-learn packages (LinearSVC and Logistic regression): 
    * https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html
    * https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html


**Case study assumptions:**
* You have access to a python installation

**Required Python libraries:**
* Pandas (https://pandas.pydata.org/)
* Numpy (www.numpy.org)
* Matplotlib (matplotlib.org)
* Scikit-learn (scikit-learn.org).

**NOTE**
* Please don't forget to save the notebook frequently when working in IPython Notebook, otherwise the changes you made can be lost.

# Read the data.

* Read the dataset with pandas and split it into train set (80%) and test set (20%)
* get the x_train, y_train (the features and the target for train set)
* get the x_test, y_test (the features and the target for test set)

In [5]:
from sklearn.svm import LinearSVC
from sklearn.datasets import make_classification
from sklearn.model_selection import cross_val_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
import pandas as pd
from sklearn.model_selection import train_test_split
data1 = pd.read_csv('data.csv')
y = data1['price_range']
X = data1.drop('price_range', axis=1)
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

## Problem 1 (30 points): Machine learning algorithms


* Examine two classifiers provided by scikit-learn 
    * LinearSVC
    * Logistic regression
* For a particular choice of parameters and classifier, look at 2 examples where the prediction was incorrect.
    * Can you conjecture on why the classifier made a mistake for this prediction?

In [4]:
lsvc = LinearSVC()
lsvc.fit(x_train, y_train)
score = lsvc.score(x_train, y_train)
print("Score: ", score)
cv_scores = cross_val_score(lsvc, x_train, y_train, cv=10)
print("CV average score:",cv_scores.mean())
y_pred = lsvc.predict(x_test)
cm = confusion_matrix(y_test, y_pred)
print("The confusion matrix is")
print(cm)
cr = classification_report(y_test, y_pred)
print("The classification report is")
print(cr)

C:\Users\User\AppData\Roaming\Python\Python310\site-packages\sklearn\svm\_base.py:1225: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


Score:  0.506875


C:\Users\User\AppData\Roaming\Python\Python310\site-packages\sklearn\svm\_base.py:1225: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\User\AppData\Roaming\Python\Python310\site-packages\sklearn\svm\_base.py:1225: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\User\AppData\Roaming\Python\Python310\site-packages\sklearn\svm\_base.py:1225: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\User\AppData\Roaming\Python\Python310\site-packages\sklearn\svm\_base.py:1225: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\User\AppData\Roaming\Python\Python310\site-packages\sklearn\svm\_base.py:1225: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\User\AppData\Roaming\Python\Python310\sit

CV average score: 0.459375
The confusion matrix is
[[ 86   0   9   0]
 [ 28   1  63   0]
 [  0   2  96   1]
 [  0   0 111   3]]
The classification report is
              precision    recall  f1-score   support

           0       0.75      0.91      0.82        95
           1       0.33      0.01      0.02        92
           2       0.34      0.97      0.51        99
           3       0.75      0.03      0.05       114

    accuracy                           0.47       400
   macro avg       0.55      0.48      0.35       400
weighted avg       0.55      0.47      0.34       400



C:\Users\User\AppData\Roaming\Python\Python310\site-packages\sklearn\svm\_base.py:1225: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


In [5]:
from sklearn.preprocessing import StandardScaler
ss= StandardScaler()
x_train= ss.fit_transform(x_train)
x_test= ss.transform(x_test)
from sklearn.linear_model import LogisticRegression
classifier=LogisticRegression(random_state=0)
classifier.fit(x_train, y_train)
y_pred= classifier.predict(x_test)
cm = confusion_matrix(y_test, y_pred)
print("The confusion Matrix is ")
print(cm)
print("The classification report is")
cr = classification_report(y_test, y_pred)
print(cr)

The confusion Matrix is 
[[ 93   2   0   0]
 [  2  85   5   0]
 [  0   3  91   5]
 [  0   0   1 113]]
The classification report is
              precision    recall  f1-score   support

           0       0.98      0.98      0.98        95
           1       0.94      0.92      0.93        92
           2       0.94      0.92      0.93        99
           3       0.96      0.99      0.97       114

    accuracy                           0.95       400
   macro avg       0.95      0.95      0.95       400
weighted avg       0.95      0.95      0.95       400



The prediction was incorrect when we trained using Linear SVC model.
The accuracy of the model for Linear SVC is 0.48 whereas The accuracy of the model for Logistic Regression in 0.95


This is beacuse Linear SVC is a classification model and our target variable price is a continous value.
Classifiers are good fit when the target variable is a discrete value such as Male or Female, True or False, Spam or Not Spam, etc.
Here price is a continous value, therefore regressors are best fit for these type of values

## Problem 2 (40 points): Use a Multi-Layer Perceptron (MLP) for classifying the costs.  Explore the parameters for the MLP and compare the accuracies against your algorithms in Problem 1.

**Read the documentation for the MLPClassifier class at https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html#sklearn.neural_network.MLPClassifier.** 
* Note: This is *very similar* to using the LinearSVC and Logistic regression classes above!
* Try different values for "hidden_layer_sizes".  What do you observe in terms of accuracy?
* Try different values for "activation". What do you observe in terms of accuracy?


In [6]:
from sklearn.neural_network import MLPClassifier
classifier = MLPClassifier(hidden_layer_sizes=(50), max_iter=300,activation = 'relu',solver='adam',random_state=1)
classifier.fit(x_train, y_train)
y_pred= classifier.predict(x_test)
cm = confusion_matrix(y_test, y_pred)
print("The confusion Matrix is ")
print(cm)
print("The classification report is")
cr = classification_report(y_test, y_pred)
print(cr)

The confusion Matrix is 
[[ 89   6   0   0]
 [  6  79   7   0]
 [  0   4  90   5]
 [  0   0   7 107]]
The classification report is
              precision    recall  f1-score   support

           0       0.94      0.94      0.94        95
           1       0.89      0.86      0.87        92
           2       0.87      0.91      0.89        99
           3       0.96      0.94      0.95       114

    accuracy                           0.91       400
   macro avg       0.91      0.91      0.91       400
weighted avg       0.91      0.91      0.91       400



C:\Users\User\AppData\Roaming\Python\Python310\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


In [7]:
from sklearn.neural_network import MLPClassifier
classifier = MLPClassifier(hidden_layer_sizes=(100), max_iter=300,activation = 'relu',solver='adam',random_state=1)
classifier.fit(x_train, y_train)
y_pred= classifier.predict(x_test)
cm = confusion_matrix(y_test, y_pred)
print("The confusion Matrix is ")
print(cm)
print("The classification report is")
cr = classification_report(y_test, y_pred)
print(cr)

The confusion Matrix is 
[[ 89   6   0   0]
 [  4  81   7   0]
 [  0   5  90   4]
 [  0   0  10 104]]
The classification report is
              precision    recall  f1-score   support

           0       0.96      0.94      0.95        95
           1       0.88      0.88      0.88        92
           2       0.84      0.91      0.87        99
           3       0.96      0.91      0.94       114

    accuracy                           0.91       400
   macro avg       0.91      0.91      0.91       400
weighted avg       0.91      0.91      0.91       400



C:\Users\User\AppData\Roaming\Python\Python310\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


In [8]:
from sklearn.neural_network import MLPClassifier
classifier = MLPClassifier(hidden_layer_sizes=(150), max_iter=300,activation = 'relu',solver='adam',random_state=1)
classifier.fit(x_train, y_train)
y_pred= classifier.predict(x_test)
cm = confusion_matrix(y_test, y_pred)
print("The confusion Matrix is ")
print(cm)
print("The classification report is")
cr = classification_report(y_test, y_pred)
print(cr)

The confusion Matrix is 
[[ 92   3   0   0]
 [  4  81   7   0]
 [  0   3  93   3]
 [  0   0   9 105]]
The classification report is
              precision    recall  f1-score   support

           0       0.96      0.97      0.96        95
           1       0.93      0.88      0.91        92
           2       0.85      0.94      0.89        99
           3       0.97      0.92      0.95       114

    accuracy                           0.93       400
   macro avg       0.93      0.93      0.93       400
weighted avg       0.93      0.93      0.93       400



C:\Users\User\AppData\Roaming\Python\Python310\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


In [9]:
from sklearn.neural_network import MLPClassifier
classifier = MLPClassifier(hidden_layer_sizes=(150), max_iter=300,activation = 'logistic',solver='adam',random_state=1)
classifier.fit(x_train, y_train)
y_pred= classifier.predict(x_test)
cm = confusion_matrix(y_test, y_pred)
print("The confusion Matrix is ")
print(cm)
print("The classification report is")
cr = classification_report(y_test, y_pred)
print(cr)

The confusion Matrix is 
[[ 94   1   0   0]
 [  2  87   3   0]
 [  0   2  92   5]
 [  0   0   1 113]]
The classification report is
              precision    recall  f1-score   support

           0       0.98      0.99      0.98        95
           1       0.97      0.95      0.96        92
           2       0.96      0.93      0.94        99
           3       0.96      0.99      0.97       114

    accuracy                           0.96       400
   macro avg       0.97      0.96      0.96       400
weighted avg       0.96      0.96      0.96       400



C:\Users\User\AppData\Roaming\Python\Python310\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


In [10]:
from sklearn.neural_network import MLPClassifier
classifier = MLPClassifier(hidden_layer_sizes=(150), max_iter=300,activation = 'tanh',solver='adam',random_state=1)
classifier.fit(x_train, y_train)
y_pred= classifier.predict(x_test)
cm = confusion_matrix(y_test, y_pred)
print("The confusion Matrix is ")
print(cm)
print("The classification report is")
cr = classification_report(y_test, y_pred)
print(cr)

The confusion Matrix is 
[[ 92   3   0   0]
 [  3  86   3   0]
 [  0   5  90   4]
 [  0   0   3 111]]
The classification report is
              precision    recall  f1-score   support

           0       0.97      0.97      0.97        95
           1       0.91      0.93      0.92        92
           2       0.94      0.91      0.92        99
           3       0.97      0.97      0.97       114

    accuracy                           0.95       400
   macro avg       0.95      0.95      0.95       400
weighted avg       0.95      0.95      0.95       400



C:\Users\User\AppData\Roaming\Python\Python310\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


In [11]:
from sklearn.neural_network import MLPClassifier
classifier = MLPClassifier(hidden_layer_sizes=(150), max_iter=300,activation = 'identity',solver='adam',random_state=1)
classifier.fit(x_train, y_train)
y_pred= classifier.predict(x_test)
cm = confusion_matrix(y_test, y_pred)
print("The confusion Matrix is ")
print(cm)
print("The classification report is")
cr = classification_report(y_test, y_pred)
print(cr)

The confusion Matrix is 
[[ 95   0   0   0]
 [  2  87   3   0]
 [  0   2  92   5]
 [  0   0   2 112]]
The classification report is
              precision    recall  f1-score   support

           0       0.98      1.00      0.99        95
           1       0.98      0.95      0.96        92
           2       0.95      0.93      0.94        99
           3       0.96      0.98      0.97       114

    accuracy                           0.96       400
   macro avg       0.97      0.96      0.96       400
weighted avg       0.96      0.96      0.96       400



C:\Users\User\AppData\Roaming\Python\Python310\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


Accuracies for Different Hidden Layers :
50 - 0.91
100- 0.91
150- 0.93

Accuarcies for different activation values:
relu - 0.93
logistic - 0.96
tanh - 0.95
identity -0.96

When compared to the accuracies of problem 1, the MLP model has a better accuracy for this dataset.
The accuracy for MLP for logistic and identity activation values with 150 as a different hidden layers value is 0.96

## Problem 3 (30 points): Running time also matters!  How fast are the algorithms versus their accuracy?
**Compare the runtime of the algorithms in Problem 1 to the running time of the MLPClassifier** 

**The jupyter command %timeit can be used to measure how long a calculation takes https://ipython.readthedocs.io/en/stable/interactive/magics.html.**
* Try different values for "hidden_layer_sizes".  What do you observe in terms of runtime?
* Try different values for "activation". What do you observe in terms of runtime?
* How long does the "fit" function take as opposed to the "predict" function?  Can you explain why?

In [22]:
%timeit lsvc = LinearSVC()

6.02 µs ± 356 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [13]:
%timeit classifier=LogisticRegression(random_state=0)

6.23 µs ± 528 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [14]:
%timeit classifier = MLPClassifier(hidden_layer_sizes=(50), max_iter=300,activation = 'relu',solver='adam',random_state=1)

21.3 µs ± 5.9 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [15]:
%timeit classifier = MLPClassifier(hidden_layer_sizes=(100), max_iter=300,activation = 'relu',solver='adam',random_state=1)

24.1 µs ± 1.31 µs per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [16]:
%timeit classifier = MLPClassifier(hidden_layer_sizes=(150), max_iter=300,activation = 'relu',solver='adam',random_state=1)

23.4 µs ± 1.22 µs per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [17]:
%timeit classifier = MLPClassifier(hidden_layer_sizes=(150), max_iter=300,activation = 'logistic',solver='adam',random_state=1)

27.4 µs ± 4.43 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [18]:
%timeit classifier = MLPClassifier(hidden_layer_sizes=(150), max_iter=300,activation = 'tanh',solver='adam',random_state=1)

24.6 µs ± 4.28 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [19]:
%timeit classifier = MLPClassifier(hidden_layer_sizes=(150), max_iter=300,activation = 'identity',solver='adam',random_state=1)

23.6 µs ± 3.11 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [20]:
%timeit classifier.fit(x_train, y_train)

C:\Users\User\AppData\Roaming\Python\Python310\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\User\AppData\Roaming\Python\Python310\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\User\AppData\Roaming\Python\Python310\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\User\AppData\Roaming\Python\Python310\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


1min 12s ± 15.4 s per loop (mean ± std. dev. of 7 runs, 1 loop each)


C:\Users\User\AppData\Roaming\Python\Python310\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


In [21]:
%timeit y_pred= classifier.predict(x_test)

18.3 ms ± 2.63 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


The run time for the models in problem 1 :
Linear SVC: 6.02 µs ± 356 ns per loop
Logistic Regression: 6.23 µs ± 528 ns per loop

As the value of hidden layers increases there is an increase in the runtime
The run time for different hidden layers is :
50 - 21.3 µs ± 5.9 µs per loop
100 - 24.1 µs ± 1.31 µs per loop
150 - 23.4 µs ± 1.22 µs per loop

The run time for different activation values are:
relu - 23.4 µs ± 1.22 µs per loop
logistic -27.4 µs ± 4.43 µs per loop 
tanh - 24.6 µs ± 4.28 µs per loop
identity - 23.6 µs ± 3.11 µs per loop

The run time for the fit function is 1min 12s ± 15.4 s per loop

The run time for the predict function is 18.3 ms ± 2.63 ms per loop

The run time of fit funtion is far more than the predict function .
This is because fit function is used to train the whole training data set.
They are also responsible for estimating the attributes out of the input data and store the model attributes and finally return the fitted estimator whereas the predict funtion will essentially use the learned parameters by fit() in order to perform predictions on new, unseen test data points.
